src/Runner/Parallel/ProcessPool.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Runner\Parallel;
+
+use React\Socket\ServerInterface;
+
+/**
+ * Represents collection of active processes that are being run in parallel.
+ * Inspired by {@see https://github.com/phpstan/phpstan-src/blob/ed68345a82992775112acc2c2bd639d1bd3a1a02/src/Parallel/ProcessPool.php}.
+ *
+ * @author Greg Korba <greg@codito.dev>
+ *
+ * @internal
+ */
+final class ProcessPool
+{
+    private ServerInterface $server;
+
+    /** @var null|(callable(): void) */
+    private $onServerClose;
+
+    /** @var array<string, Process> */
+    private array $processes = [];
+
+    /**
+     * @param null|(callable(): void) $onServerClose
+     */
+    public function __construct(ServerInterface $server, ?callable $onServerClose = null)
+    {
+        $this->server = $server;
+        $this->onServerClose = $onServerClose;
+    }
+
+    public function getProcess(ProcessIdentifier $identifier): Process
+    {
+        if (!isset($this->processes[(string) $identifier])) {
```

Comment:
```suggestion
        if (!isset($this->processes[$identifier->__toString()])) {
```
?

just to explicitly visualise that this ID is stringable and not suggest we could put any string inside `(string) $randomVar`

Reply:
I am not sure about this...  Do we actually need to know it's `Stringable`?

Reply:
you wanted to create a strong-type to replace primitive `string`, and then you take out gun and regardless what object we pass, you just cut it down to string :D 

personal preference to avoid any type-casting, eg this only triggers warning:
> var_dump((string) [""]);

You say better to avoid casting to string and keep object ID strongly typed without changes? time for `SplObjectStorage` ? :D

Reply:
But method's signature explicitly requires `ProcessIdentifier $identifier`, so it's safe to cast it to string because we're sure it's `\Stringable` 🤷‍♂️. PHPStan [ensures](https://phpstan.org/r/d70181f4-ef64-4c5f-b02c-878eb65078e3) that casting is done right, so I don't a point in calling `__toString()` directly...

Reply:
if we can have check on language compilation and not only on SCA tools, it's better.